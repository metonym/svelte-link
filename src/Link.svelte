<script context="module">
  const fetched = new Map();
</script>

<script>
  /** @restProps {a | span} */

  export let href = "javascript:void(0);";
  export let disabled = false;
  export let outbound = false;

  /** @type {string} */
  export let target = undefined;

  /** @type {string} */
  export let rel = undefined;

  async function prefetch() {
    if (fetched.has(href)) return;
    const response = await fetch(href);
    if (response.ok) fetched.set(href, true);
  }

  $: if (outbound) {
    target = "_blank";
    if (rel === undefined) rel = "noopener noreferrer";
  }
</script>

{#if disabled}
  <span
    {...$$restProps}
    on:click
    on:mouseover
    on:mouseenter
    on:mouseout
    on:focus
    on:blur
    on:keydown>
    <slot />
  </span>
{:else}
  <a
    {...$$restProps}
    {href}
    {target}
    {rel}
    on:click
    on:mouseover
    on:mouseenter
    on:mouseenter={() => {
      if (rel === 'prefetch') prefetch();
    }}
    on:mouseout
    on:focus
    on:blur
    on:keydown>
    <slot />
  </a>
{/if}
