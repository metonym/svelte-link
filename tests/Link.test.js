import { test, expect, describe, afterEach } from "vitest";
import Link from "./Link.test.svelte";

describe("Link", () => {
  let instance = null;

  afterEach(() => {
    instance?.$destroy();
    instance = null;
    document.body.innerHTML = "";
  });

  test("Link", () => {
    instance = new Link({
      target: document.body,
    });

    expect(
      document.body.querySelector("[data-basic]").outerHTML,
    ).toMatchSnapshot();
    expect(
      document.body.querySelector("[data-outbound]").outerHTML,
    ).toMatchSnapshot();
    expect(
      document.body.querySelector("[data-disabled]").outerHTML,
    ).toMatchSnapshot();
    expect(
      document.body.querySelector("[data-active]").outerHTML,
    ).toMatchSnapshot();
  });
});
