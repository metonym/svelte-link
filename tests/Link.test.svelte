<script lang="ts">
  import Link from "svelte-link";
  import Link2 from "svelte-link/Link.svelte";
</script>

<Link data-basic href="/">Text</Link>

<Link data-outbound href="https://github.com/" outbound>Text</Link>

<Link data-disabled href="/" disabled>Text</Link>

<Link data-active href="/" active>Text</Link>

<Link
  outbound
  disabled
  target="_blank"
  active
  rel=""
  href="https://github.com/"
  on:click={(e) => {
    e.preventDefault();
    console.log(e); // MouseEvent
  }}
  on:mouseover={(e) => {
    console.log(e); // MouseEvent
  }}
  on:mouseenter={(e) => {
    console.log(e); // MouseEvent
  }}
  on:mouseout={(e) => {
    console.log(e); // MouseEvent
  }}
  on:focus={(e) => {
    console.log(e); // FocusEvent
  }}
  on:blur={(e) => {
    console.log(e); // FocusEvent
  }}
  on:keydown={(e) => {
    console.log(e); // KeyboardEvent
  }}
>
  GitHub
</Link>

<Link2 outbound={false} />
