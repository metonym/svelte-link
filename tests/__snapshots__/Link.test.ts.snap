// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link > Link 1`] = `"<a data-basic="true" href="/">Text</a>"`;

exports[`Link > Link 2`] = `"<a data-outbound="true" href="https://github.com/" target="_blank" rel="noopener noreferrer">Text</a>"`;

exports[`Link > Link 3`] = `"<span data-disabled="true">Text</span>"`;

exports[`Link > Link 4`] = `"<a aria-current="page" data-active="true" href="/" class="active">Text</a>"`;
